<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#OLS-Regression" data-toc-modified-id="OLS-Regression-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>OLS Regression</a></span></li><li><span><a href="#Your-Turn----Activity-I:-Murder-rate-model" data-toc-modified-id="Your-Turn----Activity-I:-Murder-rate-model-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Your Turn -- Activity I: Murder rate model</a></span></li><li><span><a href="#Your-Turn----Activity-II:-Partial-F-Test" data-toc-modified-id="Your-Turn----Activity-II:-Partial-F-Test-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Your Turn -- Activity II: Partial F-Test</a></span></li></ul></div>

## OLS Regression

In [1]:
print('\nEnabling interactive shell outputs ...')
print('   Use command pass; to disable cell text outputs')
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

import warnings
warnings.filterwarnings('ignore') 
warnings.simplefilter(action="ignore",category=UserWarning)
warnings.simplefilter(action="ignore",category=FutureWarning)


Enabling interactive shell outputs ...
   Use command pass; to disable cell text outputs


In [2]:
import pandas as pd

vis_df = pd.read_excel('data/lect02-lin-reg.xlsx', sheet_name='Viscosity', header=0)
vis_df.head()

Temperature  Catalyst  Viscosity
0           80         8       2256
1           93         9       2340
2          100        10       2426
3           82        12       2293
4           90        11       2330

In [3]:
import statsmodels.api as sm

Y = vis_df.Viscosity
X = vis_df.drop('Viscosity', axis=1)
Xreg = sm.add_constant(X)

print(Xreg)
print(Y)

    const  Temperature  Catalyst
0     1.0           80         8
1     1.0           93         9
2     1.0          100        10
3     1.0           82        12
4     1.0           90        11
5     1.0           99         8
6     1.0           81         8
7     1.0           96        10
8     1.0           94        12
9     1.0           93        11
10    1.0           97        13
11    1.0           95        11
12    1.0          100         8
13    1.0           85        12
14    1.0           86         9
15    1.0           87        12
0     2256
1     2340
2     2426
3     2293
4     2330
5     2368
6     2250
7     2409
8     2364
9     2379
10    2440
11    2364
12    2404
13    2317
14    2309
15    2328
Name: Viscosity, dtype: int64


In [4]:
# Fit by using API
vis_lm = sm.OLS(Y, Xreg).fit()
vis_lm.params


const          1566.077771
Temperature       7.621290
Catalyst          8.584846
dtype: float64

In [5]:
# Fit by using formula
from statsmodels.formula.api import ols

formula = 'Viscosity ~ Temperature + Catalyst'
#formula = 'Viscosity ~ Temperature + Catalyst - 1'

vis_lm = ols(formula, data=vis_df).fit()
vis_lm.params


Intercept      1566.077771
Temperature       7.621290
Catalyst          8.584846
dtype: float64

In [6]:
# Making prediction
import numpy as np

Xk = X.iloc[[5],:]  # Get row 5 (as dataframe) 
Xreg_k = sm.add_constant(Xk, has_constant='add') 


Yk = np.array(vis_lm.predict(exog=Xreg_k))
Yk

array([2389.26425611])

In [7]:
print("Sum of Squared Errors (SSE): {:.2f}".format(vis_lm.ssr)) # Sum of Squares Residuals = sse
print("Sum of Squared Regression (SSR): {:.2f}".format(vis_lm.ess)) # Explained sum of squares = ssr
print("Sum of Squared Total (SST): {:.2f}".format(vis_lm.centered_tss)) # Total sum of squares = sst
print("Residual df (n-p): {:.0f}".format(vis_lm.df_resid))
print("Mean Squared errors (MSE): {:.2f}".format(vis_lm.mse_resid))
print("Standard Errors (Se): {:.2f}".format(np.sqrt(vis_lm.mse_resid)))

Sum of Squared Errors (SSE): 3478.85
Sum of Squared Regression (SSR): 44157.09
Sum of Squared Total (SST): 47635.94
Residual df (n-p): 13
Mean Squared errors (MSE): 267.60
Standard Errors (Se): 16.36


In [8]:
print("R-squared: {:.3f}".format(vis_lm.rsquared))
print("Adjusted R-squared: {:.3f}".format(vis_lm.rsquared_adj))

R-squared: 0.927
Adjusted R-squared: 0.916


## Your Turn -- Activity I: Murder rate model


In [9]:
murder_data = pd.read_excel('data/lect02-lin-reg.xlsx', sheet_name='State murder')
murder_data.describe(include='all')

Murder    Population  Illiteracy       Income       Frost
count  50.00000     50.000000   50.000000    50.000000   50.000000
mean    7.37800   4246.420000    1.170000  4435.800000  104.460000
std     3.69154   4464.491433    0.609533   614.469939   51.980848
min     1.40000    365.000000    0.500000  3098.000000    0.000000
25%     4.35000   1079.500000    0.625000  3992.750000   66.250000
50%     6.85000   2838.500000    0.950000  4519.000000  114.500000
75%    10.67500   4968.500000    1.575000  4813.500000  139.750000
max    15.10000  21198.000000    2.800000  6315.000000  188.000000

In [10]:
murder_data.head()

Murder  Population  Illiteracy  Income  Frost
0    15.1        3615         2.1    3624     20
1    11.3         365         1.5    6315    152
2     7.8        2212         1.8    4530     15
3    10.1        2110         1.9    3378     65
4    10.3       21198         1.1    5114     20

In [11]:
formula = 'Murder ~ Population + Illiteracy + Income + Frost - 1'
murder_lm = ols(formula, data=murder_data).fit()
murder_lm.params

Population    0.000224
Illiteracy    4.346314
Income        0.000251
Frost         0.002099
dtype: float64

In [12]:
print("Sum of Squared Errors (SSE): {:.2f}".format(murder_lm.ssr)) # Sum of Squares Residuals
print("Sum of Squared Regression (SSR): {:.2f}".format(murder_lm.ess)) # Explained sum of squares
print("Sum of Squared Total (SST): {:.2f}".format(murder_lm.centered_tss)) # Total sum of squares
print("Residual df (n-p): {:.0f}".format(murder_lm.df_resid))
print("Mean Squared errors (MSE): {:.2f}".format(murder_lm.mse_resid))
print("Standard Errors (Se): {:.2f}".format(np.sqrt(murder_lm.mse_resid)))

Sum of Squared Errors (SSE): 289.82
Sum of Squared Regression (SSR): 3099.67
Sum of Squared Total (SST): 667.75
Residual df (n-p): 46
Mean Squared errors (MSE): 6.30
Standard Errors (Se): 2.51


In [13]:
print("R-squared: {:.3f}".format(murder_lm.rsquared))
print("Adjusted R-squared: {:.3f}".format(murder_lm.rsquared_adj))

R-squared: 0.914
Adjusted R-squared: 0.907


In [14]:
murder_lm.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                       Results: Ordinary least squares
==============================================================================
Model:                  OLS              Adj. R-squared (uncentered): 0.907   
Dependent Variable:     Murder           AIC:                         237.7561
Date:                   2021-09-25 20:00 BIC:                         245.4042
No. Observations:       50               Log-Likelihood:              -114.88 
Df Model:               4                F-statistic:                 123.0   
Df Residuals:           46               Prob (F-statistic):          6.01e-24
R-squared (uncentered): 0.914            Scale:                       6.3005  
----------------------------------------------------------------------------------
                  Coef.      Std.Err.       t        P>|t|       [0.025     0.975]
----------------------------------------------------------------------------------
Population        0.0002       0.0001     2.4938     0.0163      0.0000     0.0004
Illiteracy        4.3463       0.5928     7.3313     0.0000      3.1530     5.5397
Income            0.0003       0.0004     0.7150     0.4782     -0.0005     0.0010
Frost             0.0021       0.0088     0.2393     0.8119     -0.0156     0.0198
------------------------------------------------------------------------------
Omnibus:                   2.407            Durbin-Watson:               2.314
Prob(Omnibus):             0.300            Jarque-Bera (JB):            1.482
Skew:                      0.345            Prob(JB):                    0.477
Kurtosis:                  3.485            Condition No.:               11821
==============================================================================
* The condition number is large (1e+04). This might indicate
strong multicollinearity or other numerical problems.
"""

## Your Turn -- Activity II: Partial F-Test

In [15]:
shear_df = pd.read_excel('data/lect02-lin-reg.xlsx', sheet_name='Strength')
shear_df.head()

Force  Power  Temperature  Time  Strength
0     30     60          175    15      26.2
1     40     60          175    15      26.3
2     30     90          175    15      39.8
3     40     90          175    15      39.7
4     30     60          225    15      38.6

In [16]:
formula = 'Strength ~ Force + Power + Temperature + Time'
shear_lm = ols(formula, data=shear_df).fit()
shear_lm.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                 Results: Ordinary least squares
=================================================================
Model:              OLS              Adj. R-squared:     0.668   
Dependent Variable: Strength         AIC:                188.0994
Date:               2021-09-25 20:00 BIC:                195.1053
No. Observations:   30               Log-Likelihood:     -89.050 
Df Model:           4                F-statistic:        15.60   
Df Residuals:       25               Prob (F-statistic): 1.59e-06
R-squared:          0.714            Scale:              26.605  
-----------------------------------------------------------------
                Coef.   Std.Err.    t    P>|t|   [0.025   0.975] 
-----------------------------------------------------------------
Intercept      -37.4767  13.0996 -2.8609 0.0084 -64.4559 -10.4974
Force            0.2117   0.2106  1.0052 0.3244  -0.2220   0.6454
Power            0.4983   0.0702  7.0997 0.0000   0.3538   0.6429
Temperature      0.1297   0.0421  3.0789 0.0050   0.0429   0.2164
Time             0.2583   0.2106  1.2268 0.2313  -0.1754   0.6920
-----------------------------------------------------------------
Omnibus:               1.712        Durbin-Watson:          2.261
Prob(Omnibus):         0.425        Jarque-Bera (JB):       1.156
Skew:                  -0.480       Prob(JB):               0.561
Kurtosis:              2.965        Condition No.:          3038 
=================================================================
* The condition number is large (3e+03). This might indicate
strong multicollinearity or other numerical problems.
"""

In [17]:
formula2 = 'Strength ~ Power + Temperature'
shear_lm2 = ols(formula2, data=shear_df).fit()
shear_lm2.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                 Results: Ordinary least squares
=================================================================
Model:              OLS              Adj. R-squared:     0.662   
Dependent Variable: Strength         AIC:                186.9755
Date:               2021-09-25 20:00 BIC:                191.1791
No. Observations:   30               Log-Likelihood:     -90.488 
Df Model:           2                F-statistic:        29.38   
Df Residuals:       27               Prob (F-statistic): 1.67e-07
R-squared:          0.685            Scale:              27.113  
-----------------------------------------------------------------
                 Coef.   Std.Err.    t    P>|t|   [0.025   0.975]
-----------------------------------------------------------------
Intercept       -24.9017  10.0721 -2.4723 0.0200 -45.5678 -4.2355
Power             0.4983   0.0709  7.0328 0.0000   0.3529  0.6437
Temperature       0.1297   0.0425  3.0499 0.0051   0.0424  0.2169
-----------------------------------------------------------------
Omnibus:               0.289        Durbin-Watson:          2.351
Prob(Omnibus):         0.866        Jarque-Bera (JB):       0.086
Skew:                  -0.127       Prob(JB):               0.958
Kurtosis:              2.935        Condition No.:          2275 
=================================================================
* The condition number is large (2e+03). This might indicate
strong multicollinearity or other numerical problems.
"""

In [18]:
from scipy import stats
result = ((shear_lm2.ssr - shear_lm.ssr)/2)/shear_lm.mse_resid
n,p=30,5; stats.f.sf(result, 2, n-p)

0.30167493019555996

##### ค่า p-value จาก partial F test = 0.30167493019555996 มีค่ามากกว่า 0.05 ดังนั้นจึง Accept H0 ก็คือตัวแปรที่เพิ่มเข้ามาไม่ช่วยการทำนาย